### PR TITLE
[CORDA-2235]: Add overrides for network parameters via command line and file

### PR DESCRIPTION
**Pull request by anthonykeenan**
Updated bootstrapper to add `--max-message-size`, `--max-transaction-size`, `--event-horizon`, which all override the default values in the network parameters file. 

Added the ability to provide a network parameters overrides file via `--network-parameter-overrides` which can be used to override the network parameters.

Removed the `--register-package-owner` and `--unregister-package-owner` options as they had some issues, now packages can be registered or unregistered using the network parameter overrides file.

image

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 1]
```diff
@@ -143,21 +143,19 @@ open class NodeStartup : NodeStartupLogging {
         val configuration = cmdLineOptions.parseConfiguration(rawConfig).doIfValid { logRawConfig(rawConfig) }.doOnErrors(::logConfigurationErrors).optional ?: return ExitCodes.FAILURE
 
         // Step 6. Configuring special serialisation requirements, i.e., bft-smart relies on Java serialization.
-        attempt { banJavaSerialisation(configuration) }.doOnException { error -> error.logAsUnexpected("Exception while configuring serialisation") } as? Try.Success
-                ?: return ExitCodes.FAILURE
+        if (attempt { banJavaSerialisation(configuration) }.doOnException { error -> error.logAsUnexpected("Exception while configuring serialisation") } !is Try.Success) return ExitCodes.FAILURE
```
Comment (anthonykeenan) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Changed these as was an IntelliJ recommendation

File: core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt [hunk 2]
```diff
@@ -217,3 +217,33 @@ data class NotaryInfo(val identity: Party, val validating: Boolean)
  * version.
  */
 class ZoneVersionTooLowException(message: String) : CordaRuntimeException(message)
+
+/**
+ * A wrapper for a legal java package. Used by the network parameters to store package ownership.
+ */
+@CordaSerializable
+data class JavaPackageName(val name: String) {
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt]: you may want to rebase as this data class has gone!
https://github.com/corda/corda/commit/504f6500226f44eda232bd460257f73c2d4a2cc6

Comment (anthonykeenan) [reply core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt]: I have rebased, I got loads of conflicts - I must have missed that one.

File: core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt [hunk 3]
```diff
@@ -217,3 +217,12 @@ data class NotaryInfo(val identity: Party, val validating: Boolean)
  * version.
  */
 class ZoneVersionTooLowException(message: String) : CordaRuntimeException(message)
+
+// Check if a string is a legal Java package name.
+private fun isPackageValid(packageName: String): Boolean = packageName.isNotEmpty() && !packageName.endsWith(".") && packageName.split(".").all { token ->
+    Character.isJavaIdentifierStart(token[0]) && token.toCharArray().drop(1).all { Character.isJavaIdentifierPart(it) }
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt]: I've removed the function, it wasn't used

File: core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt [hunk 4]
```diff
@@ -217,3 +217,12 @@ data class NotaryInfo(val identity: Party, val validating: Boolean)
  * version.
  */
 class ZoneVersionTooLowException(message: String) : CordaRuntimeException(message)
+
+// Check if a string is a legal Java package name.
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt]: I've removed the function, it wasn't used

File: docs/source/network-bootstrapper.rst [hunk 5]
```diff
@@ -247,81 +247,180 @@ To give the following:
...
+
+An example config file:
+
+.. parsed-literal::
+
+    minimumPlatformVersion=4
+    maxMessageSize=10485760
+    maxTransactionSize=524288000
+    eventHorizon="30 days"
+    packageOwnership=[
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: No it's HOCON, I've changed it, it should be the same as the node config file.

File: docs/source/network-bootstrapper.rst [hunk 6]
```diff
@@ -247,81 +247,180 @@ To give the following:
...
+        $JAVA_HOME/bin/keytool -genkeypair -keystore _teststore -storepass MyStorePassword -keyalg RSA -alias MyKeyAlias -keypass MyKeyPassword -dname "O=Alice Corp, L=Madrid, C=ES"
 
-This will generate a key store file called ``_teststore`` in the current directory.
+    This will generate a key store file called ``_teststore`` in the current directory.
 
-2. Register the package namespace to be claimed by the public key generated above:
+#. Create a ``network-parameters.conf`` file in the same directory, with the following information:
 
-.. code-block:: shell
+    .. parsed-literal::
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: Yes those whole sections were aligned incorrectly

File: docs/source/network-bootstrapper.rst [hunk 7]
```diff
@@ -247,81 +247,180 @@ To give the following:
...
+
 Package namespace ownership
-----------------------------
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Package namespace ownership is a Corda security feature that allows a compatibility zone to give ownership of parts of the Java package namespace to registered users (e.g. CorDapp development organisations).
-The exact mechanism used to claim a namespace is up to the zone operator. A typical approach would be to accept an SSL
-certificate with the domain in it as proof of domain ownership, or to accept an email from that domain.
+Package namespace ownership is a Corda security feature that allows a compatibility zone to give ownership of parts of the Java package
+namespace to registered users (e.g. CorDapp development organisations). The exact mechanism used to claim a namespace is up to the zone
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: Don't agree with the comma after e.g., but have changed it to `(e.g. a CorDapp development organisation)` which I think reads better.

File: docs/source/network-bootstrapper.rst [hunk 8]
```diff
@@ -247,81 +247,180 @@ To give the following:
...
 
-.. code-block:: shell
+    .. parsed-literal::
 
-    # Register the java package namespace using the bootstrapper tool
-    java -jar network-bootstrapper.jar --dir build/nodes --register-package-owner com.example;./_teststore;MyStorePassword;MyKeyAlias
+        packageOwnership=[
+            {
+                packageName="com.example"
+                keystore="_teststore"
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: It's a path, but defaults to being relative to the config file.

[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: I've updated the keystore description to be clear.

File: docs/source/network-bootstrapper.rst [hunk 9]
```diff
@@ -247,81 +247,180 @@ To give the following:
...
-                 [--minimum-platform-version=<minimumPlatformVersion>]
-                 [--register-package-owner java-package-namespace=keystore-file:password:alias]
-                 [--unregister-package-owner java-package-namespace]
-                 [COMMAND]
+    bootstrapper [-hvV] [--no-copy] [--dir=<dir>] [--event-horizon=<eventHorizon>]
+             [--logging-level=<loggingLevel>]
+             [--max-message-size=<maxMessageSize>]
+             [--max-transaction-size=<maxTransactionSize>]
+             [--minimum-platform-version=<minimumPlatformVersion>]
+             [-n=<networkParametersFile>] [COMMAND]
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: This is just the default output from calling `java -jar network-bootstrapper.jar --help` - I've kept it like this for consistency. I'd quite like to have the help output automatically generated and inserted into the documentation at some point but I haven't had chance to have a play around with it yet.

File: tools/bootstrapper/src/main/kotlin/net/corda/bootstrapper/NetworkParameterOverridesSpec.kt [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        Valid<NetworkParametersOverrides> = NetworkParameterOverridesSpec.parse(this, options)
+
+internal fun <T> badValue(msg: String): Valid<T> = Validated.invalid(sequenceOf(Configuration.Validation.Error.BadValue.of(msg)).toSet())
+internal fun <T> valid(value: T): Valid<T> = Validated.valid(value)
+
+internal object NetworkParameterOverridesSpec : Configuration.Specification<NetworkParametersOverrides>("DefaultNetworkParameters") {
+    private val minimumPlatformVersion by int().mapValid(::parsePositiveInteger).optional()
+    private val maxMessageSize by int().mapValid(::parsePositiveInteger).optional()
+    private val maxTransactionSize by int().mapValid(::parsePositiveInteger).optional()
+    private val packageOwnership by nested(PackageOwnershipSpec).list().optional()
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply tools/bootstrapper/src/main/kotlin/net/corda/bootstrapper/NetworkParameterOverridesSpec.kt]: No, if an empty list is supplied it means it deletes all the packages from the network parameters file, if a null is returned it just ignores the parameter.

File: docs/source/network-bootstrapper.rst [hunk 11]
```diff
@@ -247,81 +247,180 @@ To give the following:
...
+
+You can provide a network parameters overrides file using the following syntax:
+
+``java -jar network-bootstrapper-VERSION.jar --network-parameters-overrides=<filename>``
+
+Or alternatively, by using the short form version:
+
+``java -jar network-bootstrapper-VERSION.jar -n=<filename>``
+
+The network parameter overrides file is a YAML file with the following fields, all of which are optional. Any field that is not provided will be
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply docs/source/network-bootstrapper.rst]: I did!

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 12]
```diff
@@ -189,23 +192,22 @@ internal constructor(private val initSerEnv: Boolean,
...
     }
 
     private fun bootstrap(
             directory: Path,
             cordappJars: List<Path>,
             copyCordapps: Boolean,
             fromCordform: Boolean,
-            minimumPlatformVersion: Int = PLATFORM_VERSION,
-            packageOwnership: Map<String, PublicKey?> = emptyMap()
+            networkParametersOverrides: NetworkParametersOverrides = NetworkParametersOverrides()
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: ?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 13]
```diff
@@ -464,3 +455,27 @@ internal constructor(private val initSerEnv: Boolean,
...
+
+data class NetworkParametersOverrides(
+        val minimumPlatformVersion: Int? = null,
+        val maxMessageSize: Int? = null,
+        val maxTransactionSize: Int? = null,
+        val packageOwnership: List<PackageOwner>? = null,
+        val eventHorizon: Duration? = null
+)
+
+interface NetworkBootstrapperWithOverridableParameters {
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: ?

[the remark this answers is not in the document]
Comment (anthonykeenan) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt]: Yes there are other types of bootstrappers, you can enter via cordform or dockerform which have different entry points - I just created the interface so I could do some tests in the bootstrapper tool.

**Comment by anthonykeenan** [comment]
@josecoll yes I've tested them, the support for the escape characters is providied by the typesafe library natively so it didn't need any seperate unit tests in the project.

**Pull request merged**